test/controllers/declaration/SealControllerSpec.scala

```diff
@@ -0,0 +1,181 @@
+
+      val result = route(app, getRequest(uri)).value
+      status(result) must be(OK)
+    }
+
+    "populate the form fields with data from cache" in {
+      authorizedUser()
+      val request = getRequest(uri)
+      withCaching[Choice](Some(Choice(Choice.AllowedChoiceValues.StandardDec)), choiceId)
+
+      forAll(listOf[Seal](sealArbitrary.arbitrary)) { seals =>
+        withCaching[Seq[Seal]](Some(seals), Seal.formId)
+        val result = route(app, request).value
+
+        contentAsString(result).replaceCSRF mustBe
+          view(form, seals)(request).body.replaceCSRF()
+      }
+    }
+  }
+
+  ".onSubmit" should {
+
+    "return UNAUTHORIZED" when {
+
+      "user does not have an EORI" in {
+        userWithoutEori()
+        withCaching[Choice](Some(Choice(Choice.AllowedChoiceValues.StandardDec)), choiceId)
+        val body = Seq(("id", "A1"))
+        val result = route(app, postRequestFormUrlEncoded(uri, body: _*)).value
+
+        intercept[InsufficientEnrolments](status(result))
+      }
+    }
+
+    "return BAD_REQUEST" when {
+
+      "invalid data is submitted" in {
+        withCaching[Choice](Some(Choice(Choice.AllowedChoiceValues.StandardDec)), choiceId)
+
+        forAll(listOf[Seal](sealArbitrary.arbitrary)) { seals =>
+          authorizedUser()
+          withCaching[Seq[Seal]](Some(seals), Seal.formId)
+
+          val body = Seq(("id", "")) :+ addActionUrlEncoded
+          val request = postRequestFormUrlEncoded(uri, body: _*)
+
+          val result = route(app, request).value
+
+          status(result) must be(BAD_REQUEST)
+          contentAsString(result).replaceCSRF mustBe view(form.bindFromRequest()(request), seals)(request).body.replaceCSRF
+        }
+      }
+
+    }
+
+    "add seal to the cache" when {
+
+      "with valid data and on click of add" in {
+
+        forAll(arbitrary[Seal], listOf[Seal](sealArbitrary.arbitrary)) { (seal, seals) =>
+          authorizedUser()
+          withCaching[Seq[Seal]](Some(seals), Seal.formId)
+          withCaching[Choice](Some(Choice(Choice.AllowedChoiceValues.StandardDec)), choiceId)
+          val body = Seq(("id", seal.id)) :+ addActionUrlEncoded
+
+          val result = route(app, postRequestFormUrlEncoded(uri, body: _*)).value
+
+          status(result) must be(SEE_OTHER)
+          val updatedSeals = seals :+ seal
+          verify(mockCustomsCacheService)
+            .cache[Seq[Seal]](any(), ArgumentMatchers.eq(Seal.formId), ArgumentMatchers.eq(updatedSeals))(
+              any(),
+              any(),
+              any()
+            )
+        }
+      }
+    }
+    "remove seal from the cache" when {
+
+      " user click remove" in {
+
+        forAll(listOf[Seal](sealArbitrary.arbitrary)) { (seals) =>
+          whenever(seals.size > 1) {
+            authorizedUser()
+            withCaching[Seq[Seal]](Some(seals), Seal.formId)
+            withCaching[Choice](Some(Choice(Choice.AllowedChoiceValues.StandardDec)), choiceId)
```

Comment:
My propose is to use traits to define those mocks etc, it's easier to have a control on everything and to maintain the code. With this you can remove a lot of duplications

Reply:
I have updated to use in before and is much easier now